### PR TITLE
[16.0][ADD] shopinvader_api_customer

### PR DESCRIPTION
**Pull request by benwillig**
Provides a endpoint to update personal customer data

### CODE REVIEW

File: shopinvader_api_customer/routers/customer.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    partner: Annotated[ResPartner, Depends(authenticated_partner)],
+) -> Customer:
+    """
+    Get customer data of authenticated user
+    """
+    return Customer.from_res_partner(partner)
+
+
+@customer_router.post(
+    "/customer/update",
```
Comment (sebastienbeau) [inline shopinvader_api_customer/routers/customer.py]: No need of specifying the "update"

```suggestion
    "/customer",
```

**Review by sebastienbeau — CHANGES_REQUESTED** [review]
LGTM (code review) just a minor point on the route name

File: shopinvader_api_customer/routers/customer.py (hunk 1 above)
Comment (benwillig) [reply shopinvader_api_customer/routers/customer.py]: Done

File: shopinvader_api_customer/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def _update_shopinvader_customer(self, values: dict) -> "ResPartner":
```
Comment (qgroulard) [inline shopinvader_api_customer/models/res_partner.py]: We usually declare helper abstract models and make our routes depend on them. You can take inspiration from the whishlist api: https://github.com/shopinvader/odoo-shopinvader/blob/16.0/shopinvader_api_wishlist/dependencies.py

File: shopinvader_api_customer/schemas/customer.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2024 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+
+from extendable_pydantic import StrictExtendableBaseModel
+
+
+class Customer(StrictExtendableBaseModel, extra="ignore"):
```
Comment (qgroulard) [inline shopinvader_api_customer/schemas/customer.py]: I think this is a wrong idea to use the same schema for input and output. Typically, we'll want some fields to be required or hidden in one case or another.

File: shopinvader_api_customer/models/res_partner.py (hunk 2 above)
Comment (benwillig) [reply shopinvader_api_customer/models/res_partner.py]: Updated

File: shopinvader_api_customer/schemas/customer.py (hunk 3 above)
Comment (benwillig) [reply shopinvader_api_customer/schemas/customer.py]: should be fixed now

**Review by qgroulard — APPROVED** [review]

File: shopinvader_api_customer/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+This addon adds a web API on top of the res.partner model to ease the management
+of customer personal data
```
Comment (lmignon) [inline shopinvader_api_customer/readme/DESCRIPTION.rst]:
```suggestion
This addon adds a web API to allow your logged-in customer to manage their settings as its name, mobile, phone, opt-in, ...
```

File: shopinvader_api_customer/schemas/customer.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    used to get customer details
+    """
+
+    email: str
+    name: str | None = None
+    phone: str | None = None
+    mobile: str | None = None
+    opt_in: bool | None = False
+    pricelist_id: int | None = None
```
Comment (lmignon) [inline shopinvader_api_customer/schemas/customer.py:18]: IMO we should add the lang_id...

File: shopinvader_api_customer/schemas/customer.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+class CustomerUpdate(StrictExtendableBaseModel, extra="ignore"):
+    """
+    used to update customer details
+    """
+
+    name: str | None = None
+    phone: str | None = None
+    mobile: str | None = None
+    opt_in: bool | None = None
```
Comment (lmignon) [inline shopinvader_api_customer/schemas/customer.py:43]: lang_id?

File: shopinvader_api_customer/tests/test_shopinvader_customer_api.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                "property_product_pricelist": cls.pricelist.id,
+            }
+        )
+        cls.default_fastapi_authenticated_partner = cls.test_partner
+        cls.default_fastapi_router = customer_router
+
+    def _call_test_client(self, url, **kwargs):
+        method = kwargs.pop("method", "get")
+        with self._create_test_client(
+            router=self.default_fastapi_router
```
Comment (lmignon) [inline shopinvader_api_customer/tests/test_shopinvader_customer_api.py]: The `default_fastapi_router` is used by default by the method `_create_test_client`. You don't need to pass it to the method.

File: shopinvader_api_customer/tests/test_shopinvader_customer_api.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                {
+                    "opt_in": True,
+                }
+            ),
+            method="post",
+        )
+        self.assertTrue(customer.get("opt_in"))
+        self.assertFalse(partner.is_blacklisted)
+        blacklist_rec = self.Blacklist.search([("email", "=", partner.email)])
+        self.assertFalse(blacklist_rec)
```
Comment (lmignon) [inline shopinvader_api_customer/tests/test_shopinvader_customer_api.py]:
```suggestion
from odoo.tests.common import RecordCapturer


    def test_update_customer_opt_in(self):
        partner = self.test_partner
        with RecordCapture(
            self.Blacklist,
            ("email", "=", partner.email)]) as capture:

            self.assertFalse(capture.records)

            customer = self._call_test_client(
                "customer",
                content=json.dumps(
                    {
                        "opt_in": False,
                    }
                ),
                method="post",
            )
            self.assertFalse(customer.get("opt_in"))
            self.assertTrue(partner.is_blacklisted)

            self.assertTrue(capture.records)

            customer = self._call_test_client(
                "customer",
                content=json.dumps(
                    {
                        "opt_in": True,
                    }
                ),
                method="post",
            )
            self.assertTrue(customer.get("opt_in"))
            self.assertFalse(partner.is_blacklisted)
            self.assertFalse(capture.records)
```

**Review by lmignon — CHANGES_REQUESTED** [review]
Thank you for the proposal @benwillig . Some little comments but it takes a good shape.

File: shopinvader_api_customer/schemas/customer.py (hunk 5 above)
Comment (benwillig) [reply shopinvader_api_customer/schemas/customer.py:18]: I added it

File: shopinvader_api_customer/routers/customer.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+class ShopInvaderApiCustomerHelper(models.AbstractModel):
+    _name = "shopinvader_api_customer.router.helper"
+    _description = "API Customer Router Helper"
+
+    partner = fields.Many2one(
+        comodel_name="res.partner",
+    )
+
+    def _update_shopinvader_customer(self, data: CustomerUpdate) -> "ResPartner":
```
Comment (lmignon) [inline shopinvader_api_customer/routers/customer.py]:
```suggestion
    def _update_shopinvader_customer(self, data: CustomerUpdate) -> ResPartner:
```

**Review by lmignon — APPROVED** [review]

**Pull request merged**